### PR TITLE
[SPIR-V] Implement SpirvType and SpirvOpaqueType

### PR DESCRIPTION
**Pull request by cassiebeckley**
Implements hlsl-specs proposal 0011, adding `vk::SpirvType` and `vk::SpirvOpaqueType` templates which allow users to define and use SPIR-V level types.

### CODE REVIEW

File: tools/clang/lib/SPIRV/ConstEvaluator.h
Comment (s-perron) [inline tools/clang/lib/SPIRV/ConstEvaluator.h:1]: I can't tell if you have added new functionality or only moved it from SpirvEmitter. Could you make the refactoring a separate PR? This is a fairly large PR, so separating it into smaller one where possible would be nice.

**Review by s-perron — APPROVED** [review]

**Review by s-perron — CHANGES_REQUESTED** [review]
Generally look okay to me. I have a request for a few extra comments, and if you can split out the ConstEvaluator refactor into a separate PR, that would be nice.

Also @llvm-beanz should look at this as well (or someone else for MS) since you are making significant chagnes to SEMA.

File: tools/clang/lib/AST/ASTContextHLSL.cpp [hunk 1]
```diff
@@ -1241,6 +1241,26 @@ CXXRecordDecl *hlsl::DeclareNodeOrRecordType(
   return Builder.getRecordDecl();
 }
 
+#ifdef ENABLE_SPIRV_CODEGEN
+CXXRecordDecl *hlsl::DeclareInlineSpirvType(clang::ASTContext &context,
+                                            clang::DeclContext *declContext,
+                                            llvm::StringRef typeName,
+                                            bool opaque) {
+  // template<uint opcode> vk::SpirvType { ... }
```
[the remark this answers is not in the document]
Comment (cassiebeckley) [reply tools/clang/lib/AST/ASTContextHLSL.cpp:1249]: I believe I was following the format of the other methods in this file. This comment basically just shows the syntax of the type declaration that is added to the AST. I forgot to update the comment when adding size and alignment – I've now updated it.

File: tools/clang/lib/SPIRV/LowerTypeVisitor.cpp [hunk 2]
```diff
@@ -604,10 +627,115 @@ const SpirvType *LowerTypeVisitor::lowerType(QualType type,
...
+      /* PrevDecl */ nullptr);
+  QualType type = astContext.getTypeDeclType(specialized);
+
+  return type;
+}
+
+const SpirvType *LowerTypeVisitor::lowerVkTypeInVkNamespace(
+    QualType type, llvm::StringRef name, SpirvLayoutRule rule,
+    llvm::Optional<bool> isRowMajor, SourceLocation srcLoc) {
+  if (name == "SpirvType" || name == "SpirvOpaqueType") {
```
[the remark this answers is not in the document]
Comment (cassiebeckley) [reply tools/clang/lib/SPIRV/LowerTypeVisitor.cpp]: Sure, done.

**Review by s-perron — APPROVED** [review]

File: tools/clang/lib/Sema/SemaTemplate.cpp [hunk 3]
```diff
@@ -3983,8 +3983,27 @@ bool Sema::CheckTemplateArgumentList(TemplateDecl *Template,
...
+    // TODO(6246): this breaks if the frontend tries to desugar the type
+    NamedDecl *TemplatedDecl = Template->getTemplatedDecl();
+    llvm::StringRef TemplateName = TemplatedDecl->getName();
+    llvm::StringRef Namespace = "";
+    if (const auto *ND =
+            dyn_cast<NamespaceDecl>(TemplatedDecl->getDeclContext()))
+      Namespace = ND->getName();
+
+    if (Namespace != "vk" ||
+        (TemplateName != "SpirvType" && TemplateName != "SpirvOpaqueType"))
```
Comment (llvm-beanz) [inline tools/clang/lib/Sema/SemaTemplate.cpp:4001]: I'm a little nervous about this. Why is this needed? Can't we just have default values for the additional template arguments?

File: utils/hct/gen_intrin_main.txt [hunk 4]
```diff
@@ -393,6 +393,7 @@ void [[]] RawBufferStore(in u64 addr, in $funcT value);
 void [[]] RawBufferStore(in u64 addr, in $funcT value, in uint alignment);
 void [[]] ext_execution_mode(in uint mode, ...);
 void [[]] ext_execution_mode_id(in uint mode, ...);
+void [[]] ext_literal(in any<> value);
```
Comment (llvm-beanz) [inline utils/hct/gen_intrin_main.txt:396]: This also feels odd and unwieldy. Don't we know that the template parameters need to be literals, and that they need to be encoded as immediate values? Why do we need users to specify it this way?

Comment (llvm-beanz) [reply utils/hct/gen_intrin_main.txt:396]: Is that not something we should capture as a type requirement rather than an intrinsic? If we use an intrinsic to act as a cast, it seems like it should produce a value with a type annotation rather than returning void where we implicitly "look through" to the argument.

Comment (s-perron) [reply utils/hct/gen_intrin_main.txt:396]: We used the intrinsic because it already existed for this purpose.

I'm not sure what you mean by "type requirment", but we could define a new type `vk::Literal` and then cast the HLSL literal to that type. Is that what you are thinking about?

File: tools/clang/lib/Sema/SemaTemplate.cpp (hunk 3 above)
Comment (cassiebeckley) [reply tools/clang/lib/Sema/SemaTemplate.cpp:4001]: There's currently no way to represent `vk::SpirvType` in HLSL syntax or in the AST. The closest representation would probably be something like

```hlsl
template <uint OpCode, uint size, uint alignment, typename... args>
struct SpirvType {}
```

However, that would only allow us to pass types to `SpirvType`, whereas we would like to be able to pass either types or values of any type. So instead this PR defines `SpirvType` as

```hlsl
template <uint OpCode, uint size, uint alignment>
struct SpirvType {}
```

and modifies Sema to ignore any additional arguments. This actually doesn't work for every situation where someone might reasonably want to use `SpirvType`, so I've created #6246 as a follow-up issue.

The options to fix it will likely be either to add `typename... args` to the `SpirvType` template and specifically disable validation for that parameter pack, or to add a new type of parameter pack which will allow `SpirvType` to be accurately represented in the AST.

**Comment by cassiebeckley** [comment]
@sudonatalie @llvm-beanz I've updated this PR based on the design changes to 0011.

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl [hunk 5]
```diff
@@ -0,0 +1,39 @@
+// RUN: %dxc -T ps_6_0 -E main -fcgl  %s -spirv | FileCheck %s
```
Comment (sudonatalie) [inline tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl:1]: Did you mean to clean up the TODOs in this test?

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.hlsl [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// RUN: %dxc -T ps_6_0 -E main -fcgl  %s -spirv | FileCheck %s
+
+// TODO(5751): enable Array test when `using` has been implemented
```
Comment (sudonatalie) [inline tools/clang/test/CodeGenSPIRV/spv.inline.type.hlsl:3]: #5751 is closed now, can these be uncommented?

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl (hunk 5 above)
Comment (cassiebeckley) [reply tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl:1]: Hm, this is something we should support once #5554 is fixed. Do you think it would be better to commit this with the TODOs, or to simply remove the test altogether?

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.hlsl (hunk 6 above)
Comment (cassiebeckley) [reply tools/clang/test/CodeGenSPIRV/spv.inline.type.hlsl:3]: Hm, this still doesn't work completely. Something to do with how Texture2D is something of a special case, I think.

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl (hunk 5 above)
Comment (sudonatalie) [reply tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl:1]: Hm, that's unfortunate. If we're going to keep the test in this PR then let's leave the TODO but uncomment the test itself and make it expected to fail, so that if/when #5554 is fixed it will get caught for an update.

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.hlsl (hunk 6 above)
Comment (cassiebeckley) [reply tools/clang/test/CodeGenSPIRV/spv.inline.type.hlsl:3]: I'm *really* not sure what's causing this. It seems to be an error from `SemaHLSL.cpp`, but it doesn't happen for non-SpirvType code (https://godbolt.org/z/KeoGMaEqs).

I've filed a bug at #6498. I'll reference the bug in this test.

File: tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl (hunk 5 above)
Comment (cassiebeckley) [reply tools/clang/test/CodeGenSPIRV/spv.inline.type.enum-class.hlsl]: Done.

**Comment by sudonatalie** [comment]
LGTM

@llvm-beanz Can you take another look?

**Review by llvm-beanz — APPROVED** [review]
Sorry for the delay in reviewing this. It looks great to me. Thank you!
